src/basic/virt.c

```diff
@@ -195,19 +195,34 @@ static int detect_vm_dmi(void) {
 
 static int detect_vm_xen(void) {
         _cleanup_free_ char *domcap = NULL;
+        int r;
+
+        r = access("/proc/xen/capabilities", F_OK);
+        if (r < 0) {
+                if (r == -ENOENT)
```

Comment:
access() returns error codes in errno, not in its return value. This must be changed to "if (errno == ENOENT)" hence.

That said, access() with F_OK checks existance anyway, there are very few other reasons why this might fail. Hence I'd really just write this as:

```c
return access("…", F_OK) < 0 ? VIRTUALIZATION_NONE : VIRTUALIZATION_XEN;
```